driver/src/main/java/org/neo4j/driver/internal/connector/socket/SSLSocketChannel.java

```diff
@@ -0,0 +1,455 @@
+     * greater than the size of deciphered data.
+     * Otherwise, the deciphered bytes or the bytes that could not fit into {@code buffer} would be left in {@code
+     * plainIn} buffer.
+     *
+     * If the byes in {@code plaintIn} will not be used outside this method, we should always clear {@code
+     * plainIn} after each call to avoid wasting memory on it.
+     * <p>
+     * When working with this method, we should always put this method in a loop as there is no guarantee
+     * that deciphering would be carried out successfully in each call. The possible reasons that deciphering fails
+     * in a call could be requiring more bytes to decipher, too small buffers to hold all data, etc.
+     *
+     * To verify if deciphering is done successfully, we could check if any bytes has been read into {@code buffer},
+     * as the deciphered bytes will only be saved into {@code buffer} when deciphering is carried out successfully.
+     *
+     * @param buffer
+     * @return
+     * @throws IOException
+     */
+    private HandshakeStatus unwrap( ByteBuffer buffer ) throws IOException
+    {
+        HandshakeStatus handshakeStatus;
+
+        if ( channel.read( cipherIn ) < 0 )
+        {
+            throw new ClientException( "Failed to establish SSL socket connection." );
+        }
+        cipherIn.flip();
+
+        Status status = null;
+        do
+        {
+            status = sslEngine.unwrap( cipherIn, plainIn ).getStatus();
+            // Possible status here:
+            // OK - good
+            // BUFFER_UNDERFLOW - we need to read more data to carry out unwrap
+            // BUFFER_OVERFLOW - we need to enlarge plainIn to hold all deciphered data
+            if ( status == OK )
+            {
+                plainIn.flip();
+                bufferCopy( plainIn, buffer );
+                plainIn.compact();
+            }
+            else if ( status == BUFFER_OVERFLOW ||
+                      (status == BUFFER_UNDERFLOW && cipherIn.limit() == cipherIn.capacity()) )
+            {
+                // Enlarge the app buffer and re-enter this method
+                int oldSize = plainIn.capacity();
+                if ( oldSize >= bufferMax )
+                {
+                    // TODO
+                    throw new ClientException( "Failed to enlarge a buffer: plainIn.size = " + oldSize +
+                                               ", cipherIn.size = " + cipherIn.capacity() +
+                                               ". The maximum size for a buffer is " + bufferMax );
+                }
+                int newSize = Math.min( oldSize * 2, bufferMax );
+                ByteBuffer newPlainIn = ByteBuffer.allocateDirect( newSize );
+                bufferCopy( plainIn, newPlainIn );
+                plainIn = newPlainIn;
+                logger.debug( "Enlarged application input buffer from " + oldSize + " to " + newSize +
```

Comment:
Since this is in a super low-level loop of the driver, could we do an explicit check for isDebugEnabled here before we construct the string? Otherwise we'll create a stringBuilder, chararray and a string object twice every time a network packet is pulled in